### ISSUE TITLE
How to use WebApi / Serilog with .Net 4.7.2

### ISSUE THREAD

**Issue by fjoppe-brunel** [thepirat000/Audit.NET#622]
Bom Dia,

I am trying to implement WebApi for an existing code base, which requires auditing, but I do not see any logs.

My setup is:
* **.Net 4.7.2**
* F# programming language
* `System.Web.Http.ApiController`
* Serilog + Serilog data provider

 I have played around with F# script and the following "works":
```fsharp
#r "nuget: Audit.NET.Serilog, 21.0.3"
#r "nuget: Audit.NET, 21.0.3"
#r "nuget: Serilog, 3.0.1"
#r "nuget: Serilog.Sinks.Console, 4.1.0"


open Audit.Core
open Audit.NET.Serilog
open Serilog

let log = 
    (new LoggerConfiguration())
        .WriteTo.Console()
        .CreateLogger()

Audit.Core.Configuration.Setup()
    .UseSerilog(fun config ->
        config.Logger(fun ev -> log.ForContext(System.Type.GetType(ev.EventType)))
            .LogLevel(fun ev -> if ev.Environment.Exception <> null then LogLevel.Fatal else LogLevel.Info)
            .Message(fun auditEvent -> auditEvent.ToJson() |> box)
            |> ignore
    )

Audit.Core.AuditScope.Log("Admin:ChangeUser", {| Test = "test1" |})

Audit.Core.AuditScope.Log("Admin:ChangeUser",  {| Test = "test2" |})
```

And "works" means, I get output which I can further refine.

The next step failed, in which I wanted to set it up for Rest Api's. Also later, I want to set it up for MVC, which is used by the web user interface.

I have setup a serilog logger using a file sink, which is a configuration copy of a working debug log.

In the global.asax code behind, I have the following:

```fsharp
    member this.InitAudit() =
        Audit.Core.Configuration.Setup()
            .UseSerilog(fun config ->
                config.Logger(fun ev -> << return the serilog logger >>)
                    .LogLevel(fun ev -> if ev.Environment.Exception <> null then LogLevel.Fatal else LogLevel.Info)
                    .Message(fun auditEvent -> auditEvent.ToJson() |> box)
                    |> ignore
            )

    member this.Application_Start() =
        this.InitAudit() |> ignore
```

And on top of the class I wanted to test, I've added the `AuditApi` attribute:
```fsharp
[<AuditApi(EventTypeName = "{controller}/{action} ({verb})", IncludeResponseBody = true, IncludeRequestBody = true, IncludeModelState = true)>]
[<RoutePrefix("api/admin")>]
type MyAuditedController(someGetter) =
    inherit ApiController(someGetter) 

    [<HttpGet>]
    [<Route("some-function")>]
    member this.SomeFunction() = 
        this.operation()
```

I've also checked the loglevel, which is set to Verbose in every location.

Did I forget something? The logfile isn't even created.

**Comment by thepirat000** [comment]
At first sight, your code looks good.

Are you sure that the `Audit.Core.Configuration.Setup().UseSerilog()` is ever called? 
If for any reason it's not, Audit.NET will be generating .json files as logs in the app folder by default. Do you see any file being created when you call your API action?

**Comment by fjoppe-brunel** [comment]
Yes, the UseSerilog() is called - breakpointed it. 

Also, if I remove this Serilog initialization, nothing is generated in folder "WebApp/bin". 

Note, I am running in Debug mode from Visual Studio.

**Comment by thepirat000** [comment]
What package and version are you referencing for the WebApi audit? `Audit.WebApi` or `Audit.WebApi.Core` ?

You should be referencing `Audit.WebApi`.

Can you try to do `open Audit.WebApi` `let scope = this.GetCurrentAuditScope()` inside your action method to see if the audit scope is being created?

**Comment by fjoppe-brunel** [comment]
Hi Frederico,

I only use `Audit.WebApi`, I am not using dotnetcore anywhere.

I have inserted the scope assignment you suggested above. Breakpointed it and the Audit.Net scope is returned.
![image](https://github.com/thepirat000/Audit.NET/assets/120374642/af570781-6c9a-45d9-a392-4c2c14db18b7)

It looks like as everything has been initialized correctly.

I also connected the Debug logger to the Audit logger, and then I got logging in the correct logfile.

Could the issue be because this is an async function?

![image](https://github.com/thepirat000/Audit.NET/assets/120374642/0f0f6aca-530f-4e61-a07a-1d75f03a1433)

**Comment by thepirat000** [comment]
Can you share a minimal runnable project that reproduces the issue so I can help further?

I was unable to reproduce and had too little experience with F#.

**Comment by fjoppe-brunel** [comment]
I have continued with testing - have not make a repro yet, I'll try this later. 

My findings:
* Calling `Audit.Core.AuditScope.Log` in an Api method, produces logging;
* Installing package `Audit.Mvc` (21.0.4) and setting the `[Audit]` attribute on a class inheriting from `System.Web.Mvc.Controller`, does produce audit logging (nice!);
* Attribute `[AuditApi]` does not work on a class inheriting from `System.Web.Http.ApiController`, nor does the `[Audit]` attribute (ok, just tryin :) );

Also, the controllers in the (legacy) code have a multi-layered inheritence scheme as follows: 
`A` inherits `B` inherits `System.Web.Mvc.Controller`

The `[Audit]` tag worked on class `B`, while not on class `A`. The `[AuditApi]` tag didn't work at all.

**Comment by fjoppe-brunel** [comment]
Unfortunately, I was unable to create a simple project, because the legacy code base is very big, and stripping it just caused other issues. So I took the other approach - get source code of `Audit.WebApi` into my solution and temporary use this instead of the nuget package, for investigation.

After setting the attribute and adding a few breakpoints:
- [This line](https://github.com/thepirat000/Audit.NET/blob/497cce1723e4be9be6c6260cf766bf0b50b3e3b8/src/Audit.WebApi/AuditApiAdapter.cs#L45) was hit, while
- [This line](https://github.com/thepirat000/Audit.NET/blob/497cce1723e4be9be6c6260cf766bf0b50b3e3b8/src/Audit.WebApi/AuditApiAdapter.cs#L91) was never hit;

Meaning that the attribute works. I've surrounded [this line](https://github.com/thepirat000/Audit.NET/blob/497cce1723e4be9be6c6260cf766bf0b50b3e3b8/src/Audit.WebApi/AuditApiAttribute.cs#L99) with a `try`/`catch` all, and it always came into the `catch` section.

It always received the exception that binary `System.Net.Http.Formatting.dll` version 5.2.6.0 could not be found/loaded. [This Stack Overflow response](https://stackoverflow.com/a/65785609) helped me further.

My code was using `Microsoft.AspNet.WebApi.Client` version **5.2.7.0**, and related libraries too. There also was a [version redirection](Microsoft.AspNet.WebApi.Client) in my `Web.Config`, to version **5.2.7.0**.

**Solution:**
After I downgraded every related dependency to "5.2.6.0", including the version redirect, Audit.WebApi is working!
